### PR TITLE
Shutdown should not block on exceptions from targets

### LINKED ISSUE #3958: NLog Target that blocks in FlushAync can never be closed
This makes it impossible to complete `LogManager.Shutdown()`

It will correctly complain with timeout error on Flush using threadpool-threads. But afterwards when closing all targets then it will get stuck in Target.Close because there threadpool-thread is holding SyncRoot inside FlushAsync.

Ofcourse a NLog Target should not perform blocking operations inside `FlushAsync` and hold the SyncRoot-lock forever, but it could be nice that  `LogManager.Shutdown()` could still complete (with alerts) and not be stuck forever.

Maybe have a check that if the Flush-completes with timeout (because of missing cal to continuation), then it should also validate that all threadpool-threads have exited the FlushAync-method of the target. If threadpool-threads are stuck, then it should perform async-close-operation (or just skip target-close) instead of blocking shutdown forever.

### PR DESCRIPTION
**Pull request by snakefoot**
LogFactory Shutdown performs async close of configuration, when timeout on flush

Trying to resolve #3958

### CODE REVIEW

File: src/NLog/LogFactory.cs [hunk 1]
```diff
@@ -254,8 +254,19 @@ public LoggingConfiguration Configuration
                     if (oldConfig != null)
                     {
                         InternalLogger.Info("Closing old configuration.");
-                        Flush();
-                        oldConfig.Close();
+                        if (!FlushAllTargetsSync(oldConfig, DefaultFlushTimeout, ThrowExceptions))
+                        {
+                            var manualResetEvent = new ManualResetEvent(false);
+                            AsyncHelpers.ForEachItemInParallel(new[] { oldConfig }, ex => manualResetEvent.Set(), (old, cont) => { old.Close(); cont(null); });
```
Comment (304NotModified) [inline src/NLog/LogFactory.cs:260]: Maybe we could make a new method in AsyncHelpers? `ForEachItemInParallel(new[] { oldConfig }` is a bit strange to read. 

(PS I prefer also to break this line with some newlines, e.g. something like this:

```c#
 AsyncHelpers.ForEachItemInParallel(new[] { oldConfig },
     ex => manualResetEvent.Set(), 
     (old, cont) => { old.Close(); cont(null); }
   );

```

File: src/NLog/LogFactory.cs [hunk 2]
```diff
@@ -254,8 +254,19 @@ public LoggingConfiguration Configuration
                     if (oldConfig != null)
                     {
                         InternalLogger.Info("Closing old configuration.");
-                        Flush();
-                        oldConfig.Close();
+                        if (!FlushAllTargetsSync(oldConfig, DefaultFlushTimeout, ThrowExceptions))
```
Comment (304NotModified) [inline src/NLog/LogFactory.cs:257]: Funny that the setter of Configuration has changed. it expected to see some changes in Flush(async) or Shutdown methods. 

Could this PR have more side effects?

File: src/NLog/LogFactory.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/LogFactory.cs]: Added newlines.

File: src/NLog/LogFactory.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/LogFactory.cs:257]: Well it has the side-effect that if flush fails with an exception, then it uses a threadpool thread for closing the configuration. And will continue after 15 secs (Even if a target requires 2 minutes to flush and close).

**Comment by 304NotModified** [comment]
@snakefoot Do you think it;s wise to release this one with NLog 4.7.2?

**Comment by snakefoot** [comment]
> @snakefoot Do you think it;s wise to release this one with NLog 4.7.2?

I think something should be done about this blocking forever. Since it points fingers at NLog for not working. And I also think 30 seconds is reasonable long time to wait for flush and close.

Since the async closing of targets only happens when flush has failed with exception, then I was thinking it was safe.

**Comment by snakefoot** [comment]
Think this PR needs some more thought, since unsure if there will be issues with holding LogFactory-lock and doing async Close on other thread.

Actually think the flush should be done outside the lock, and it should also do the async-close outside the lock if the flush fails. Maybe better for NLog5.

**Comment by snakefoot** [comment]
Inactive for too long

**Pull request closed without merging**